### ISSUE TITLE
Anomaly detection on text fields

### ISSUE THREAD

**Issue by srinivaskasturi** [oracle/tribuo#394]
**Ask the question**
Can we have an example for Anomaly detection on text fields.

I'm trying to use anomaly detection on a text field. I'm loading the training and test data using a JsonDataSource

My training data is
[
  {"id": 1, "type": "A"},
  {"id": 2, "type": "B"},
  {"id": 3, "type": "A"},
  {"id": 4, "type": "B"}
]

and using a single field processor for "type" field: 
fieldProcessors.put("type", new TextFieldProcessor("type", new BasicPipeline(new BreakIteratorTokenizer(Locale.US),2)));

My test data is
[
{"id": 1, "type": "C"}
]

Instead of flagging it as anomalous, i get the below exception
java.lang.IllegalArgumentException: No features found in Example ArrayExample(numFeatures=1,output=EXPECTED,weight=1.0,features=[(type@1-N=C, 1.0)])

Note: I have set the FieldResponseProcessor default value to EXPECTED in both test and training data as leaving it as unknown in test data gives me the below exception
java.lang.IllegalArgumentException: The sentinel Unknown Output was used as a ground truth label in example number 0

**System details**
- Tribuo version: 4.3.1
- Java version: 11
- OS/Architecture: Mac OS

**Comment by Craigacp** [comment]
The basic pipeline you are using will emit a single feature with value 1 for each token emitted by the tokenizer. When you trained the model it saw features "A" and "B", and so built a model with 2 features. The test example only contains feature "C", which the trained model doesn't know and so it is removed then you get the "no features found in example" `IllegalArgumentException`. If you want it to detect anomalous text of this kind then you'll need to embed the text using something like the `BERTFeatureExtractor`, but that embedding approach won't work very well with single words (or characters). If you have single words and you want to detect anomalous words which have different meanings from the training data, then you could consider using a trained word2vec model and using the embedding for the word as the feature vector in Tribuo.

For the second `IllegalArgumentException` is that coming out of the anomaly evaluation system, or the `model.predict()` call? The former is expected (though possibly a little unhelpful), but the latter is a bug.

**Comment by srinivaskasturi** [comment]
Thanks Craig. That helps!
The second IllegalArgumentException was coming out of the anomalyEvaluator.evaluate(model, dataset) call. So I guess it is expected. Closing the issue now.
